url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/resid/Filter.java

```diff
@@ -0,0 +1,1250 @@
+
+	protected int f0_count;
+
+	// ----------------------------------------------------------------------------
+	// Inline functions.
+	// The following functions are defined inline because they are called every
+	// time a sample is calculated.
+	// ----------------------------------------------------------------------------
+
+	/**
+	 * SID clocking - 1 cycle
+	 * 
+	 * @param voice1
+	 * @param voice2
+	 * @param voice3
+	 * @param ext_in
+	 */
+	public void clock(int /* sound_sample */voice1,
+			int /* sound_sample */voice2, int /* sound_sample */voice3,
+			int /* sound_sample */ext_in) {
+		// Scale each voice down from 20 to 13 bits.
+		voice1 >>= 7;
+		voice2 >>= 7;
+
+		// NB! Voice 3 is not silenced by voice3off if it is routed through
+		// the filter.
+		if ((voice3off != 0) && ((filt & 0x04) == 0)) {
+			voice3 = 0;
+		} else {
+			voice3 >>= 7;
+		}
+
+		ext_in >>= 7;
+
+		// This is handy for testing.
+		if (!enabled) {
+			Vnf = voice1 + voice2 + voice3 + ext_in;
+			Vhp = Vbp = Vlp = 0;
+			return;
+		}
+
+		int /* sound_sample */Vi = Vnf = 0;
+		// Route voices into or around filter.
+
+		if (ANTTI_LANKILA_PATCH) {
+
+			if ((filt & 1) != 0)
+				Vi += voice1;
+			else
+				Vnf += voice1;
+			if ((filt & 2) != 0)
+				Vi += voice2;
+			else
+				Vnf += voice2;
+			if ((filt & 4) != 0)
+				Vi += voice3;
+			else
+				Vnf += voice3;
+			if ((filt & 8) != 0)
+				Vi += ext_in;
+			else
+				Vnf += ext_in;
+		} else {
+			// The code below is expanded to a switch for faster execution.
+			// (filt1 ? Vi : Vnf) += voice1;
+			// (filt2 ? Vi : Vnf) += voice2;
+			// (filt3 ? Vi : Vnf) += voice3;
+
+			switch (filt) {
+			default:
+			case 0x0:
+				Vi = 0;
+				Vnf = voice1 + voice2 + voice3 + ext_in;
+				break;
+			case 0x1:
+				Vi = voice1;
+				Vnf = voice2 + voice3 + ext_in;
+				break;
+			case 0x2:
+				Vi = voice2;
+				Vnf = voice1 + voice3 + ext_in;
+				break;
+			case 0x3:
+				Vi = voice1 + voice2;
+				Vnf = voice3 + ext_in;
+				break;
+			case 0x4:
+				Vi = voice3;
+				Vnf = voice1 + voice2 + ext_in;
+				break;
+			case 0x5:
+				Vi = voice1 + voice3;
+				Vnf = voice2 + ext_in;
+				break;
+			case 0x6:
+				Vi = voice2 + voice3;
+				Vnf = voice1 + ext_in;
+				break;
+			case 0x7:
+				Vi = voice1 + voice2 + voice3;
+				Vnf = ext_in;
+				break;
+			case 0x8:
+				Vi = ext_in;
+				Vnf = voice1 + voice2 + voice3;
+				break;
+			case 0x9:
+				Vi = voice1 + ext_in;
+				Vnf = voice2 + voice3;
+				break;
+			case 0xa:
+				Vi = voice2 + ext_in;
+				Vnf = voice1 + voice3;
+				break;
+			case 0xb:
+				Vi = voice1 + voice2 + ext_in;
+				Vnf = voice3;
+				break;
+			case 0xc:
+				Vi = voice3 + ext_in;
+				Vnf = voice1 + voice2;
+				break;
+			case 0xd:
+				Vi = voice1 + voice3 + ext_in;
+				Vnf = voice2;
+				break;
+			case 0xe:
+				Vi = voice2 + voice3 + ext_in;
+				Vnf = voice1;
+				break;
+			case 0xf:
+				Vi = voice1 + voice2 + voice3 + ext_in;
+				Vnf = 0;
+				break;
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 392:465, 553:619.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.